duracloud/duracloud
# Duracloud 1191: Adds support for range requests and automatic retrying of midstream failures from point of failure.

**Adds support for range requests and automatic retrying of midstream failures from point of failure.**
* * *

**JIRA Ticket**:https://jira.duraspace.org/browse/DURACLOUD-1191

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)
https://jira.duraspace.org/browse/DURACLOUD-1208

# What does this Pull Request do?
Adds support in the storeclient for performing range requests on content streams.
    Additionally now streams are automatically resumed from a point of failure in the case of
    an IOException encountered in midstream.
    
    All users of the storeclient (including the retrievaltool) will automatically benefit from
    this improvement as momentarily network failures will not trigger redownloading of an entire
    file (be it a chunk or otherwise).

# How should this be tested?
There are unit and integration tests covering the functionality. 

Additionally you could start retrieving a large file  (say 100 MB) and watch the file on disk as it approached completion.  Before it completes turn off wifi, and then turn it back on.  You should see the file continue to completion rather than starting over.

# Additional Notes:
Any additional information that you think would be helpful when reviewing this PR.

Example:
* Does this change require documentation to be updated? Yes.
* Does this change add any new dependencies? No. 
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? No
* Could this change impact execution of existing code? Yes

# Interested parties
Tag (@ mention) interested parties or, if unsure, @duracloud/committers
